mist/worker/src/main/scala/io/hydrosphere/mist/worker/WorkerActor.scala

```diff
@@ -15,37 +16,53 @@ import scala.util.{Failure, Success}
 
 case class ExecutionUnit(
   requester: ActorRef,
-  jobFuture: Future[Either[String, Map[String, Any]]]
+  jobFuture: Future[Map[String, Any]]
 )
 
-trait JobStarting { that: Actor =>
-
+trait JobStarting {
+  that: Actor =>
+  val runnerSelector: RunnerSelector
   val namedContext: NamedContext
-  val runner: JobRunner
+  val artifactDownloader: ArtifactDownloader
 
-  def startJob(req: RunJobRequest)(implicit ec: ExecutionContext): Future[Either[String, Map[String, Any]]] = {
+  protected final def startJob(req: RunJobRequest)(implicit ec: ExecutionContext): Future[Map[String, Any]] = {
     val id = req.id
-    val future = Future {
-      namedContext.sparkContext.setJobGroup(req.id, req.id)
-      runner.run(req, namedContext)
-    }
+    sender() ! JobFileDownloading(id)
+    val jobStart = for {
+      file   <- artifactDownloader.downloadArtifact(req.params.filePath)
+      runner =  runnerSelector.selectRunner(file)
+      res    <- runJob(req, runner)
+    } yield res
```

Comment:
`JobStarted` doesn't actually wait for downloading, it will fire right after `JobFileDownloading`